### ISSUE TITLE
[BUG] Incorrect behavior of asynchronous D2D copy

### ISSUE THREAD

**Issue by jeng1220** [cupy/cupy#2159]
`cupy.copyto` misses to apply user specified stream.
Please see following mini repro.

* Conditions
  - CuPy 5.4.0 and 6.0.0rc1
  - Ubuntu 16.04
  - CUDA 10.1
* Code to reproduce
```python
import cupy as cp

def main():
  SIZE = 250000000
  a = cp.empty((1, SIZE), cp.float32)
  b = cp.empty((1, SIZE), cp.float32)
  my_stream = cp.cuda.stream.Stream()

  my_stream.use()
  cp.copyto(a, b) # <- it will use default stream but it should use `my_stream`
  my_stream.synchronize()

  my_stream.use()
  a[:,:] = b # <- use `my_stream`
  my_stream.synchronize()

if __name__ == '__main__':
  main()
```
* Error messages, stack traces, or logs
[nvvp log](https://drive.google.com/file/d/1cc2bkKVTBk60pQVXW0log7BsLK6-RFY2/view?usp=sharing)

**Comment by jeng1220** [comment]
Hi,
I just found another issue, which was very similar above one.
* Conditions
CuPy 5.4.0 and 6.0.0rc1
Ubuntu 16.04
CUDA 10.1

* Code to reproduce
```python
import cupy as cp

def main():
  SIZE = 10000
  my_stream = cp.cuda.stream.Stream()
  my_stream.use()
  a = cp.empty((SIZE, SIZE), cp.float32)
  b = cp.empty((SIZE, SIZE), cp.float32)
  # `cp.dot` includes a small async-H2D-copy with default stream at first time,
  # but it should use `my_stream`
  c = cp.dot(a, b) 
  my_stream.synchronize()

if __name__ == "__main__":
  main()
```
* Error messages, stack traces, or logs
[nvvp log 2](https://drive.google.com/file/d/1fF9pNglpDjHbvAMWVMg9J4JMQjfqDkfp/view?usp=sharing)

**Comment by takagi** [comment]
I made a PR to fix the `cupy.copyto` issue. As for `cupy.dot`, the H2D copy is triggered by initializing CUBLAS. With the fact it is fairly tiny, ~1us or so, I suppose it is not an issue to address.

**Comment by takagi** [comment]
I profiled the `cupy.dot` reproduction code above with nvprof and here is the screenshot viewed on nvvp. The highlighted small yellow bars are the H2D copy @jeng1220 reported and the green long bars on right are `sgemm` that actually computes dot production.

This H2D copy seems to be called by CUBLAS library in its initialization, `cublasCreate`. As CuPy lazily initializes CUBLAS when it uses CUBLAS handle for the first time, when CUBLAS is initialized is not clear for users.

<img width="1312" alt="スクリーンショット 2020-01-14 15 18 51" src="https://user-images.githubusercontent.com/1224375/72319263-13bec680-36e2-11ea-89d8-4b6b092cad8c.png">

**Comment by takagi** [comment]
@nboshi Would you take a look at https://github.com/cupy/cupy/issues/2159#issuecomment-574027347 ?

**Comment by niboshi** [comment]
@takagi 
Thanks.
If the synchronization only occurs once at the first `dot`, we can ignore it.
The situation is similar to other functions that trigger kernel compilation on the fly (loading a CUDA module incurs synchronization too.).

**Comment by niboshi** [comment]
Closing, as `cupy.copyto` has been fixed in #2942 .
@jeng1220 @takagi Thanks!
